File: mkdocs/overrides/devbook/websockets/listen-transaction-error.mjs
const NODE_URL = process.env.NODE_URL
	|| 'https://reference.symboltest.net:3001';
const WS_URL = NODE_URL.replace('http', 'ws') + '/ws';
console.log(`Using node ${NODE_URL}`);

const MONITOR_ADDRESS = process.env.MONITOR_ADDRESS
	|| 'TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I';
console.log(`Monitoring address: ${MONITOR_ADDRESS}`);

const websocket = new WebSocket(WS_URL);

// Connect and receive uid
const uid = await new Promise((resolve) => {
	websocket.addEventListener('message', (event) => {
		const msg = JSON.parse(event.data);
		resolve(msg.uid);
	}, { once: true });
});
console.log(`Connected to ${WS_URL} with uid ${uid}`);

// Subscribe to status channel
const channel = `status/${MONITOR_ADDRESS}`;
websocket.send(JSON.stringify({uid, subscribe: channel}));
console.log('Subscribed to status channel');

// Handle incoming messages
websocket.addEventListener('message', (event) => {
	const msg = JSON.parse(event.data);
	const txHash = msg.data.hash;
	const code = msg.data.code;
	console.log(
		`Transaction ${txHash.substring(0, 16)}... `
		+ `rejected with code: ${code}`
	);
});

// Unsubscribe on exit
process.on('SIGINT', () => {
	websocket.send(JSON.stringify({uid, unsubscribe: channel}));
	console.log('Unsubscribed from status channel');
	websocket.close();
	process.exit(0);
});

File: mkdocs/overrides/devbook/websockets/listen-transaction-error.py
import asyncio
import json
import os

from websockets import connect

NODE_URL = os.getenv('NODE_URL', 'https://reference.symboltest.net:3001')
WS_URL = NODE_URL.replace('http', 'ws', 1) + '/ws'
print(f'Using node {NODE_URL}')

MONITOR_ADDRESS = os.getenv(
	'MONITOR_ADDRESS',
	'TCHBDENCLKEBILBPWP3JPB2XNY64OE7PYHHE32I'
)
print(f'Monitoring address: {MONITOR_ADDRESS}')


async def main():
	async with connect(WS_URL) as websocket:
		# Connect and receive uid
		response = json.loads(await websocket.recv())
		uid = response['uid']
		print(f'Connected to {WS_URL} with uid {uid}')

		# Subscribe to status channel
		channel = f'status/{MONITOR_ADDRESS}'
		await websocket.send(json.dumps(
			{'uid': uid, 'subscribe': channel}
		))
		print('Subscribed to status channel')

		# Handle incoming messages
		try:
			async for raw_message in websocket:
				msg = json.loads(raw_message)
				tx_hash = msg['data']['hash']
				code = msg['data']['code']
				print(
					f'Transaction {tx_hash[:16]}... '
					f'rejected with code: {code}'
				)

		# Unsubscribe on exit
		finally:
			await websocket.send(json.dumps(
				{'uid': uid, 'unsubscribe': channel}
			))
			print('Unsubscribed from status channel')


try:
	asyncio.run(main())
except KeyboardInterrupt:
	pass
except Exception as error:
	print(error)
